### ISSUE TITLE
Regex Match, Split and Matches should support RegexOptions.EnvironmentNewLine as (?=[\r\n]|\z)

### ISSUE THREAD

**Issue by jzabroski** [dotnet/corefx#28410]
From [MSDN](https://docs.microsoft.com/en-us/dotnet/standard/base-types/regular-expression-options#Multiline):

> By default, $ matches only the end of the input string. If you specify the RegexOptions.Multiline option, it matches either the newline character (\n) or the end of the input string. It does not, however, match the carriage return/line feed character combination. To successfully match them, use the subexpression \r?$ instead of just $.

This is by far one of the biggest gotchas with using .NET Regex class. I suggest adding a RegexOptions.EnvironmentNewLine which treats $ as matching Environment.NewLine - it can be not set by default, even though this is almost ALWAYS what EVERYONE wants!

**Comment by jnm2** [comment]
Ohhhh.... 😮

**Comment by svick** [comment]
I don't like that this would depend on `Environment.NewLine`. Transferring files between different OSes is common, so I think there is a good chance you will need to process files with different line endings than those that are native to your OS. Instead, I think a setting that would match any line ending (independent of the current OS) would be better. It could be called something like `AnyNewLine`.

Here is a table comparing the options (✓ means processing files with the given line-ending style works well on the given OS, ✗ means it does not):

| OS | Line-ending style | Current | `EnvironmentNewLine` | `AnyNewLine` |
|:-:|:-:|:-:|:-:|:-:|
| Windows | Windows | ✗ | ✓ | ✓ |
| Windows | Unix | ✓ | ✗ | ✓ |
| Unix | Windows | ✗ | ✗ | ✓ |
| Unix | Unix | ✓ | ✓ | ✓ |

**Comment by jzabroski** [comment]
Petr, that is exactly what I meant if you looked at how I defined it:
(?=[\r\n|\z) can match any mix of line endings, even files trashed by mixed
line endings. AnyNewline is a MUCH better name than what I came up with.
Love teamwork!

Adding a portable sigil for Environment.NewLine probably isn't a bad idea,
but the only pragmatic way I could think of doing it would be to export all
Regex sigils and let the consumer of the API reprogram the sigil mapping. I
thought about this, but felt it was too complicated and also ruin the
tooling around current .NET regular expressions. Sites like RegexHero make
writing Regex so easy that I wanted a solution that would be easy for tools
to adopt.

The only gotchas I typically encounter are forgetting to escape a pipe, and
then forgetting about how line ending semantics works. Tools help me with
the former, but the latter always trips me up.

**Comment by jzabroski** [comment]
Actually, I see now that in my original description my description of the behavior is inconsistent. Thank you for that wonderful table. I will update the first post with it.

**Comment by jzabroski** [comment]
@jnm2 @svick  I updated the original description. Please give feedback.

**Comment by ViktorHofer** [comment]
I like the idea and its cross-platform benefit.

> I think a setting that would match any line ending

How is _any line ending_ defined? The name suggests that it can handle *any* line ending, are there more than Unix and Windows line ending chars? We could probably refine the name.

**Comment by svick** [comment]
@ViktorHofer 

> How is any line ending defined?

According to [Wikipedia's summary of the Unicode standard](https://en.wikipedia.org/wiki/Newline#Unicode), there are 8 character sequences that form a line break (7 single characters, including CR and LF, and the sequence CR+LF).

Though I think those other characters are not widely used, so supporting only either LF and CR+LF (Unix and Windows) or CR, LF and CR+LF (old Mac, Unix and Windows) would be fine, especially if it lead to better performance than fully supporting the Unicode definition of line break.

@jzabroski

> (?=[\r\n]|\z)

I don't understand this regex, as far as I can tell, it wouldn't match `\r\n` (and so `\r\n` would be reported as two line breaks).

**Comment by jzabroski** [comment]
I'll write some test cases with XUnit and use that as a specification. We can then tweak the test cases to come up with a final specification/acceptance criteria. [Fare](https://github.com/moodmosaic/Fare) could also be used to confirm.

**Comment by svick** [comment]
@terrajobst 

> The table looks wrong (Windows on Windows on the Current should work IMHO)

I don't think the table is wrong, see [the documentation](https://docs.microsoft.com/en-us/dotnet/standard/base-types/regular-expression-options#Multiline) linked in the top post, where it says that you should use `\r?$` to match Windows-style line endings in multiline mode.

You can also try running code like the following:

```c#
var match = Regex.Match("foo\r\nbar", ".*$", RegexOptions.Multiline);

foreach (char c in match.Value)
    Console.WriteLine($"{c} (0x{(int)c:X2})");
```

For me, running it on .Net Core 2.1 RTM on Windows 10 prints:

```
f (0x66)
o (0x6F)
o (0x6F)
 (0x0D)
```

While if this was "working well", I would expect to get just the three characters of `"foo"` as the match, not `"foo\r"`.

**Comment by jzabroski** [comment]
@terrajobst  I agree with @svick . Intuitively, all his table is saying is that neither $ nor Environment.NewLine is fully portable solution for the two ubiquitous platforms.  AnyNewLine therefore proposes to fill that gap.

**Comment by jzabroski** [comment]
@terrajobst 

I'll update the original post to modify the enum options to be 0x0400 (1024)

**Comment by Mpdreamz** [comment]
Taking a stab at this for the Hackaton

**Comment by ViktorHofer** [comment]
Ok, assigned you.

**Comment by Mpdreamz** [comment]
Had some fun this afternoon going through the internals of `System.Text.RegularExpressions`

I pushed what I have so far here: https://github.com/Mpdreamz/corefx/commit/4c4d6882802feae0a50e19dd18033ab6af648270

I might pick it up for fun at some point again, feel free to unassign me 😄

**Comment by ViktorHofer** [comment]
wow that's great progress! it's up to you if you want to continue working on it or pass it on to me / other community members.

**Comment by danmosemsft** [comment]
Unassigning for someone else to look. Thanks for the start!

**Comment by shishirchawla** [comment]
Hi @danmosemsft, I would like to have a go at this. Can you please get this assigned to me. Thanks.

**Comment by shishirchawla** [comment]
Hi @danmosemsft / @jzabroski  ,

Heres my PR for the new API - https://github.com/dotnet/corefx/pull/41195 . This is my first PR on the project, so please bare with me. 

Looking forward to your feedback.

Shishir

**Comment by jzabroski** [comment]
@danmosemsft @terrajobst @ViktorHofer @svick  It seems that during the process of implementing this feature, @shishirchawla discovered an area to improve the API a tiny bit further.  If you go to the Anchors documentation, there are two anchors, `$` and `\Z`, which mention Windows-specific `\n` line ending.  I've updated the top post of this issue with the full notes, including @terrajobst API review notes, so that everything is consolidated in one post.

I like @shishirchawla 's improvement, but I think the final focus should be on how this is communicated in the Docs, such as [https://docs.microsoft.com/en-us/dotnet/standard/base-types/anchors-in-regular-expressions](https://docs.microsoft.com/en-us/dotnet/standard/base-types/anchors-in-regular-expressions) and ensuring full test coverage.

**Comment by danmosemsft** [comment]
@terrajobst could you please give your take on the `\Z` change? Does this need to go back to review or is it arguably what was intended?

**Comment by jzabroski** [comment]
@danmosemsft  So sorry I'm a foot dragger on this.  Been trying to keep up with my own open source project, life, work, etc.

Relevant examples / test cases:

https://docs.microsoft.com/en-us/dotnet/standard/base-types/anchors-in-regular-expressions#end-of-string-or-line-
https://docs.microsoft.com/en-us/dotnet/standard/base-types/anchors-in-regular-expressions#end-of-string-or-before-ending-newline-z
https://docs.microsoft.com/en-us/dotnet/standard/base-types/anchors-in-regular-expressions#end-of-string-only-z

The examples are a bit idiosyncratic in that they reference `Environment.NewLine`, and it probably isn't immediately clear to the documentation writer that the example output will change depending on the runtime OS environment.

**Comment by danmosemsft** [comment]
ping @terrajobst

**Comment by terrajobst** [comment]
The name `AnyNewLine` works for me. Unless someone of @dotnet/fxdc disagress, I wouldn't bring this back to API review. Our feedback was mostly around semantics, and that seems to have been take care of.

**Comment by danmosemsft** [comment]
The question wasn't about `AnyNewLine` (that was what was originally approved) but confirming that you expected the new behavior for `\Z` as well as `$`. I cannot imagine you don't...

**Comment by jzabroski** [comment]
I still owe you all some sample documentation, I think, too... been stuck on various other projects and _yak shaving_.

**Comment by danmosemsft** [comment]
@jzabroski @shishirchawla , @terrajobst remark is sufficient for me. We can assume `\Z` is desired as well.

**Comment by jzabroski** [comment]
@danmosemsft Great - @shishirchawla still needs to update the PR with the review comments.  I do think we want to explicitly cover the test case given by @svick  earlier, as its _exactly_ the scenario that confuses everyone.  My head hurts every time I try to predict the right answer after not looking at this thread for a week.

**Comment by jzabroski** [comment]
@danmosemsft @shishirchawla Updated the this Issue Title to reflect reality.
